### PR TITLE
UHM-6029 issues with date of test results on Add Past Labs

### PR DESCRIPTION
**Pull request by brandones**
https://pihemr.atlassian.net/browse/UHM-6029

Peek 2022-01-05 18-53

### CODE REVIEW

File: configuration/pih/htmlforms/labResults.xml [hunk 1]
```diff
@@ -175,19 +168,34 @@
...
+                </div>
+            </div>
+            <div>
+                <label for="specimen-collection-date">
+                    <uimessage code="pihcore.lab.specimen_collection_date"/>
+                </label>
+                <div class="small required">
+                    <obs id="specimen-collection-date"
+                         conceptId="PIH:SPUTUM COLLECTION DATE"
+                         allowTime="false" />
```
Comment (mogoodrich) [inline configuration/pih/htmlforms/labResults.xml:191]: It looks like you are adding a new obs here to store sputum collection date?  Previously, we had decided that specimen collection date would be tracked as the encounter date... for better or worse, as this form is in use, we can't change it without performing a data migration.  (fwiw, the idea was that the encounter date should get set to the date when the other obs in the encounter (ie the results) are clinically-relevant, which is the specimen collection data)

There's some general server-side validation that doesn't allow any of the obs value datetimes on a form to be after the encounter date, which is the source of the current error (this validation may not have been working previously which is why this is just coming up now).  We should disable the server-side validation for this, via the "allowFutureDates" feature, see: https://wiki.openmrs.org/display/docs/HTML+Form+Entry+Module+Reference#HTMLFormEntryModuleReference-%3Cobs%3E

Besides that, having JS logic to update the test date if necessary when the specimen date changes and vice versa makes sense to me,  thought  might be easier to just have a submission handler that validates that test date >= encounter date upon submission.  This is documented here, see "Defining javascript functions to call before form validation and submission":

https://wiki.openmrs.org/display/docs/HTML+Form+Entry+JavaScript+Reference

This not documented well/at all, but it's also possible to have the one-question-per-screen navigator jump back to a specific field when you display this.  I forget the details, but there are some API methods provided here, and it looks like there are some examples of usage within the form itself.

https://github.com/openmrs/openmrs-module-uicommons/blob/master/omod/src/main/webapp/resources/scripts/navigator/navigator.js#L119

**Review by mogoodrich — CHANGES_REQUESTED** [review]
Thanks Brandon, see my comment on the ticket.

**Comment by brandones** [comment]
Thanks for the thorough explanation @mogoodrich . I think I've fixed it.

Peek 2022-01-06 16-31

File: configuration/pih/htmlforms/labResults.xml [hunk 2]
```diff
@@ -104,8 +104,14 @@
...
 
             jq(function() {
-                getField('specimen-collection-date.value').datepicker('option', 'maxDate', new Date());
+                getField('encounter-date.value').change(function(evt) {
+                    setValue('specimen-collection-date.value', evt.target.value);
+                    getField('test-date.value').datepicker('option', 'minDate', evt.target.value);
+                });
+                setValue('specimen-collection-date.value', new Date());
+                getField('encounter-date.value').datepicker('option', 'maxDate', new Date());
                 getField('test-date.value').datepicker('option', 'maxDate', new Date());
```
Comment (mogoodrich) [inline configuration/pih/htmlforms/labResults.xml:113]: Doesn't this line need to be removed?

File: configuration/pih/htmlforms/labResults.xml [hunk 3]
```diff
@@ -104,8 +104,14 @@
...
+                getField('test-date.value').datepicker('option', 'minDate', new Date());
```
Comment (mogoodrich) [inline configuration/pih/htmlforms/labResults.xml:114]: Does this work properly during edit? ie, if you are are opening an encounter that has a both a encounter date and a test date in the past?

**Review by mogoodrich — COMMENTED** [review]
Added a few questions, thanks @brandones !

File: configuration/pih/htmlforms/labResults.xml [hunk 4]
```diff
@@ -175,19 +169,35 @@
...
+                <label for="specimen-collection-date">
+                    <uimessage code="pihcore.lab.specimen_collection_date"/>
+                </label>
+                <div class="small required">
+                    <encounterDate id="encounter-date" default="now" />
+                    <div style="display: none;">
+                        <obs id="specimen-collection-date"
+                            conceptId="PIH:SPUTUM COLLECTION DATE"
+                            allowTime="false" />
+                    </div>
```
[the remark this answers is not in the document]
Comment (dmdesimone) [reply configuration/pih/htmlforms/labResults.xml:184]: this works for me.  would like @lnball to weigh in also.

File: configuration/pih/htmlforms/labResults.xml [hunk 5]
```diff
@@ -104,8 +104,14 @@
             }
 
             jq(function() {
-                getField('specimen-collection-date.value').datepicker('option', 'maxDate', new Date());
+                getField('encounter-date.value').change(function(evt) {
+                    setValue('specimen-collection-date.value', evt.target.value);
+                    getField('test-date.value').datepicker('option', 'minDate', evt.target.value);
+                });
+                setValue('specimen-collection-date.value', new Date());
+                getField('encounter-date.value').datepicker('option', 'maxDate', new Date());
```
Comment (brandones) [inline configuration/pih/htmlforms/labResults.xml:112]:
```suggestion
```

File: configuration/pih/htmlforms/labResults.xml (hunk 2 above)
Comment (brandones) [reply configuration/pih/htmlforms/labResults.xml:113]: Good catch, thanks

File: configuration/pih/htmlforms/labResults.xml (hunk 3 above)
Comment (brandones) [reply configuration/pih/htmlforms/labResults.xml]: It didn't, but it does now

File: configuration/pih/htmlforms/labResults.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (brandones) [reply configuration/pih/htmlforms/labResults.xml]: I've changed this. I had thought I'd seen the `setValue` function in HFE working, but it now seems that it doesn't work at all, at least with date-type obs. There's no more specimen collection date obs. It's just the encounter date.

**Review by mogoodrich — APPROVED** [review]
LGTM, thanks @brandones !

**Pull request merged**